### DISCUSSION TITLE
Should side-by-side head collisions be evaluated as head-to-head collisions?

### DISCUSSION THREAD

**Discussion by bcambl** [BattlesnakeOfficial/feedback#162] [Q&A] [answered]
In the following game at [Turn 160](https://play.battlesnake.com/g/2051ea6f-be5a-4406-8f3f-adacb7ff471d/?turn=160) there is a head-to-head collision where the snake heads involved are right next to each other: https://play.battlesnake.com/g/2051ea6f-be5a-4406-8f3f-adacb7ff471d/

In this situation both snakes clipped into each other resulting in a body collision and a double elimination despite orange snake being longer.

@coreyja mentioned that this was discussed briefly on a recent stream with @bvanvugt: 
at timestamp 1:08 https://www.twitch.tv/battlesnakeofficial/video/1277883598

![Kapture 2022-01-29 at 15 00 31](https://user-images.githubusercontent.com/1531419/151680335-9923ced5-da15-4db6-8a4a-a58d7a9117a2.gif)
_(edited by @bvanvugt to add a gif showing what 'side-by-side' head collisions look like)_

**Comment by bvanvugt** [comment]
Yeah this is an interesting side effect of playing wrapped games on a game board with odd dimensions. Technically, the collision is working as currently intended - both snakes move into each others bodies and are eliminated via body collision.

There are 3 options we're currently weighing, and I'd be interested in everyone's feedback:
* Move competitive wrapped games to a board with even dimensions (ie: 10x10). This prevents the scenario from happening altogether.
* Detect this scenario specifically and apply head-to-head collision rules rather than body collision rules.
* Leave as is, and have players consider it in their algorithm.

One of the reasons we launched the global wrapped arena was specifically to observe instances of this and see what the community thinks :) We also have a 10x10 version ready to test, that we might switch to next week as a trial.

**Reply by bcambl** [reply]
Based on the 3 proposed options, I think having head-to-head rules applied to the situation would be the best solution as it seems reasonable and would cover all game modes and board dimensions going forward.

Moving competitive wrapped games to even dimensions would be a great workaround depending on the effort  required to update and test the rules prior to the next league.

Would leaving this _as is_ conflict with the existing collision logic where the blue snake moves right or down (fully surrounded via wrapped or non-wrapped) and is eliminated first?

**Reply by chillfox** [reply]
I prefer to either make it a head to head or leave it as is over changeing the board dimensions.

**Reply by ambadoom** [reply]
I'm not a fan of making It head-to-head as it would be a bit of a special case (Where two heads have collided but they are not on the same square). Having both snakes die feels like more of a consequence of the existing rules. I would be for changing the board dimensions to remove the confusion entirely

**Reply by bcambl** [reply]
The more I think about this.. the more I think simply changing the board dimensions +/- makes the most sense and requires no change to the rules/code.
@chillfox, any particular reason you are opposed to an adjustment to board dimensions?

**Reply by mawif** [reply]
Changing board dimensions also needs the starting positions to be asymmetric, right? (So they're offset by a odd number of cells in both "normal" and "wrapped" directions.) Does that mess with the balance changes to do with initial food positioning, also mentioned during the stream?

**Reply by chillfox** [reply]
@bcambl yeah, it would break my wrapped handling code.
Having odd board dimensions neatly allows handling wrapped mode without having to make all of the snake logic wrapped aware.
So I calculate an offset for placing the snake in the centre of the board. Then apply that offset to food, hazard and other snakes. This then allows my pathfinding and flood fill algorithms to not be aware of board wrapping.
With even board dimensions there is no single centre square.

**Reply by smallsco** [reply]
Heh, I remember having this exact discussion back in 2018, and we settled on the second option at that time, updating the old engine to treat head-on-neck collisions as if they were head-to-head, so there is historical precedent :)

I'd be fine with either doing it again, or doing nothing at all. The "do nothing" option is somewhat interesting as it adds an additional strategic element to the Wrapped game mode: if you want that head-on-head kill you have to work extra hard to get it.

Not a fan of switching to a board with even dimensions, as others have mentioned it affects initial food placement as there is no centre square.

**Reply by JonathanArns** [reply]
I agree with the sentiment that changing board dimensions is not the best solution. Two main reasons:

1. It basically says that odd board sizes are incorrect for wrapped and disables playing wrapped on half the board sizes (or at least the "correct" wrapped rules). It's worth considering that battlesnake exists outside the official league and maybe people would like to play on different boards elsewhere.

2. A lot of snakes (but not all) will break by changing board size. My snake for example really only likes 11x11 and smaller, this thread also already contains another example. It feels slightly "unfair" to change the rules in a fundamental way that only affects some developers. The other rule change of adding special behaviour would affect every developer, but would probably not break any snakes completely.

Personally I kind of like the added complexity to the game that comes from doing nothing here, but we can really only see with time if it actually makes things more interesting or just more weird. Especially for new developers though I find that kind of complexity unreasonably weird compared to the rest of battlesnake. With that in mind I think that the small rule change of adding special behaviour for this situation is the best solution, since the game still works on all boards as intended and as expected and it probably won't break any snakes completely.

**Reply by Hegberg** [reply]
Personally I don't like the idea of changing it to a head to head collision, since you are trying to move your snake into an opposing snakes old head position, not actually colliding with the opposing snakes head current tile, so the snake didn't calculate the collision properly.

I think that changing it to add a special rule behaviour is more detrimental than helpful for new developers, because instead of just "body collisions always result in death" you have a fine print "but not in these oddly specific situations for this 1 game mode".

I also highly value simple concrete rule sets without exceptions that allow for depth and nuance when you start getting into specific situations, and this is exactly one of those scenarios.

I would vote for currently doing nothing this spring league, and seeing how it plays out with this rule set. Nobody really has too much experience with this situation yet, and it's difficult to judge whether changing this behaviour is needed and/or it is a positive or negative change to make to the game.

**Reply by BoldBigflank** [reply]
Option 2 is pretty flawed because snakes can choose whether to go into that scenario.
If two snakes are side by side, the smaller one would never want to move toward the other snake.
If we allowed snakes to bite the necks of other snakes, then when two snakes are side by side the bigger can choose to attack the smaller for a safe and unavoidable kill.

**Reply by BoldBigflank** [reply]
I'm splitting out my reasoning for my favorite since it's difficult to parse these responses in this thread.

**Reply by JonathanArns** [reply]
@BoldBigflank I think that's not true. The way this would be implemented, if I understand correctly, is that these situations get counted as a head to head collision, if and only if both snakes land on the neck of each other. So if a bigger snake walks onto the neck of a smaller snake, while the smaller snake walks to an empty space, the bigger snake loses because it is a body collision.

@Hegberg I would be fine with that as well. Just waiting and maybe making a change in summer if needed seems like a pretty safe option.

**Comment by BoldBigflank** [comment]
There's an option that hasn't been suggested yet:

What if the edges were kind of a tunnel, where it takes 2 turns to go from one side of the screen to the other. The snake wouldn't have a choice to make while traveling, it would always continue through to the other side.

This will let you keep an odd-sized board, and head-to-heads will stay as they are now.

**Reply by bvanvugt** [reply]
That's an interesting idea 🤔

**Reply by JonathanArns** [reply]
That's a very cool idea. My immediate thought is that the time delay might be a big disadvantage for snakes going through, since their enemies basically know their next move in advance.

**Reply by BoldBigflank** [reply]
The downside to this is it forces you to have an odd sized arena. It's the same issue as option 1 above, in a sense.

There's also the added complexity to describing the state of the board to the players' snakes. Should they be told the state when they have no choice of direction? How would players' algorithms have to be adjusted when some of their moves will span 2 turns?

**Reply by JonathanArns** [reply]
I thought about it a little more and I realized that this option is actually the same a changing the board size while introducing a new kind of field that doesn't allow snakes to change their direction.
Only changing the board size seems like a strictly better option to me now.

**Comment by jlafayette** [comment]
> Detect this scenario specifically and apply head-to-head collision rules rather than body collision rules.

This is my favorite solution because when watching the game, this is what I expect to happen and it just feels right to me.

I'm ok with either of the other options too though, but I would prefer to leave it as-is rather than changing board dimensions.  I like how the starting position on odd boards has one food per snake and then one food directly in the middle.

**Reply by BoldBigflank** [reply]
Copying my notes on that option for relevant discussion:
> Option 2 is pretty flawed because snakes can choose whether to go into that scenario.
If two snakes are side by side, the smaller one would never want to move toward the other snake.
If we allowed snakes to bite the necks of other snakes, then when two snakes are side by side the bigger can choose to attack the smaller for a safe and unavoidable kill.

**Reply by JonathanArns** [reply]
@BoldBigflank Quoting my reply from other discussion :wink: 
> The way this would be implemented, if I understand correctly, is that these situations get counted as a head to head collision, if and only if both snakes land on the neck of each other. So if a bigger snake walks onto the neck of a smaller snake, while the smaller snake walks to an empty space, the bigger snake loses because it is a body collision.

**Reply by BoldBigflank** [reply]
Right, if it's only when both move into each other's necks, a smaller snake has absolutely no reason to move into the neck of a bigger snake and can easily choose to move into another direction, including into itself, so that it won't give a win to a bigger snake.

**Reply by JonathanArns** [reply]
Well how is that different from normal head to head collisions then?
A smaller snake can avoid those just the same if it has space to dodge. And if it doesn't, it either kills itself or the collision happens, in both rulesets.

**Reply by BoldBigflank** [reply]
Because in the normal, 2 tile difference scenario, the bigger snake survives if the smaller snake decides to dodge.
In this situation, the bigger snake will die hitting the body of the smaller snake.

**Reply by jlafayette** [reply]
After watching some games and playing in the wrapped league for a while, I actually would like to revise my opinion to prefer option ~~2~~ 3:
> Leave as is, and have players consider it in their algorithm.

What's nice about this is that the normal logic of avoiding colliding with another snakes body seems to work well in this new situation too, and so it'll be one less special case to program for.

**Reply by BoldBigflank** [reply]
Cool, but that's option 3 :)

**Comment by BoldBigflank** [comment]
> Leave as is, and have players consider it in their algorithm.

I've played just a few games with it as is and I think I like this one the best. I understand this undermines the creator's original design forcing snakes to be an even number of spaces apart, but this was an unspoken rule in the first place. The game isn't broken when snakes are 1 tile apart, there's just fewer opportunities to have head-to-head collisions.

**Reply by JonathanArns** [reply]
It might also lead to some interesting situations where a loosing snake could force a tie.

**Comment by bvanvugt** [comment]
Just wanted to say thanks to everyone for the feedback. It means a lot that y'all care so much and can be so constructive about this 🙏 

We're going to parse through it this week, clarify our own thinking, and make a call in the next few days. We'll update this discussion accordingly.

**Comment by bvanvugt** [comment, marked as answer]
Okay! Thanks again everyone for the constructive feedback, it's much appreciated 🤗 

**For spring league we've decided to leave this scenario as-is and encourage developers to account for it in their algorithms.** After Spring League concludes, we may revisit this decision in favor of counting it as a head-to-head collision. We will not be changing the board dimensions to address this scenario.

A lot of excellent points were made in this thread, and I wanted to detail some of our thinking...

Changing board dimensions won't solve the problem long-term, and we decided we want to be clear and decisive with where we stand on this scenario in a broader sense - not just for competitive wrapped play. Switching the board dimensions will also cause heavy disruption to the starting game state. We're excited to experiment with the starting state, but feel this is not a great reason to start doing so (especially so close to the start of a competitive league).

It's also important to note that an existing Battlesnake will still perform reasonably well in this situation when left as-is. A naive avoidance algorithm will steer away from side-by-side heads, and so will a naive look-ahead algorithm. Advanced algorithms can gain an advantage by being aware of this scenario - specifically, a smaller snake could actively choose to wrap knowing that it becomes immune to head-to-head elimination - and that's the sort of situation we like to promote. Those combined are pretty compelling to leave as-is for Spring League.

That said, we want to acknowledge the original post by @bcambl (and comment from @jlafayette) mentioning that, as an observer, this scenario FEELS like a head-to-head collision and anything else is counter-intuitive. This is super important to note, and it wasn't lost on us. Obviousness and accessibility for spectators are especially important when it comes to welcoming new folks to the community. Battlesnake games should be fun for everyone, not just participants, so the observer experience matters a lot to us. If we're going to revisit this decision later this year, this would be the primary motivation.

If you have additional thoughts/questions, post here or you can always find Chris and me in Discord :)

**Reply by bvanvugt** [reply]
Just a quick follow up on this - we revisited this issue internally and with the community after Spring League and we're going to keep the existing behavior for the remainder of the 2022 Competitive Season 🏆 👍
